virttest/graphical_console.py

```diff
@@ -229,6 +415,95 @@ def _key_event(self, down, keys):
             events.append(event)
         self._vm.qmp_monitors[0].input_send_event(events)
 
+    def _btn_event(self, event, btn):
+        """
+        Send mouse button event.
+
+        :param event: Button event, defined as True=down, False=up.
+        :param btn: Mouse button.
+        """
+
+        events = [{"type": "btn",
+                   "data": {"down": event,
+                            "button": btn}}]
+        self._vm.monitor.input_send_event(events)
+
+    def _scroll_event(self, event):
+        """
+        Send mouse scroll event.
+
+        :param event: Event as vertical up/down or horizontal left/right.
+                      But qmp only support scroll vertical up/down.
+                      Does not support scroll horizontal left/right currently.
+        """
+
+        if event == "up":
+            self._btn_event(self.BTN_DOWN, self.SCROLL_FORWARD)
+        elif event == "down":
+            self._btn_event(self.BTN_DOWN, self.SCROLL_BACKWARD)
+        else:
+            raise ConsoleNotSupportedError("qmp console doesn't support"
+                                           "scroll %s" % event)
+
+    def _motion_event(self, pos, mtype):
+        """
+        Send mouse motion event from qmp monitor.
+
+        :param pos: Mouse position as (x, y).
+        :param mtype: Mouse move type, with abs for absolute coordinates,
+                      with rel for relatively coordinates.
+        """
+
+        x, y = self._translate_pos_qmp(pos, mtype)
+        events = [{"type": mtype,
+                   "data": {"axis": "x",
+                            "value": int(x)}},
+                  {"type": mtype,
+                   "data": {"axis": "y",
+                            "value": int(y)}}]
+        self._vm.monitor.input_send_event(events)
+        self.pointer_pos = pos
+
+    def _translate_pos_qmp(self, pos, mtype):
+        """
+        Translate position coordinates to qmp recognized coordinates.
+
+        :param pos: Mouse position value as a tuple.
+                    the pos value range in console as: x in 0 ~ screen width,
+                    y in 0 ~ screen height.
+        :param mtype: Mouse move type, with abs for absolute coordinates,
+                      with rel for relatively coordinates.
+
+        :return: A tuple that translated position (x, y) for qmp.
+        """
+
+        if mtype == "abs":
+            x = math.ceil(pos[0] * 32767 / self.width)
+            y = math.ceil(pos[1] * 32767 / self.height)
+        else:
+            x = pos[0] - self.pointer_pos[0]
+            y = pos[1] - self.pointer_pos[1]
+        return (x, y)
+
+    def screen_size(self):
+        """
+        Get the current screen size in pixels.
+        Returns a tuple of 2 integers
+        """
+
+        image_file = "/tmp/screendump"
+        self._vm.monitor.screendump(image_file)
+        img = Image.open(image_file)
+        return img.size
```

Comment:
Instead of directly manipulating the image, it is better to introduce a function (i.e. `image_size`) to the `ppm_utils` module and call it at here
```python
def image_size(filename):
    img = Image.open(image_file)
    return img.size
```

Reply:
thanks, will update~